core/txflow/src/dag/mod.rs

```diff
@@ -24,17 +26,22 @@ pub struct DAG<'a, P: 'a + Payload, W: 'a + WitnessSelector> {
 
     witness_selector: &'a W,
     starting_epoch: u64,
+
+    misbehaviour: &'a mut M,
```

Comment:
This is not going to work if we report misbehavior on multiple levels of our infra, since there can only be one mutable reference at a time. I suggest we wrap it into &'a RefCell<M>, since the reporting of misbehaviors is very short and atomic so it is easy to make sure that at the runtime we do not have two misbehaviors reported at the same time (which will cause RefCell to panic, because it essentially moves the borrowing checks from compile-time to runtime).

Later our Misbehavior reporter will be performing heavy tasks and which would involve composing the proof of the misbehavior and so we would want it to be in a separate process. Then it might become more complex, but for now let's not think about it at simply print/save the misbehavior reports.